contract/models/contract.py

```diff
@@ -217,12 +212,61 @@ def action_contract_send(self):
             'context': ctx,
         }
 
+    @api.model
+    def _get_recurring_create_invoice_domain(self, contract=False):
+        domain = []
+        date_ref = fields.Date.context_today(self)
+        if contract:
+            contract.ensure_one()
+            date_ref = contract.recurring_next_date
+            domain.append(('id', '=', contract.id))
+        domain.extend(
+            [
+                ('recurring_invoices', '=', True),
+                ('recurring_next_date', '<=', date_ref),
+            ]
+        )
+        return domain
+
+    @api.multi
+    def _get_lines_to_invoice(self, date_ref=False):
+        self.ensure_one()
+        if not date_ref:
+            date_ref = fields.Date.context_today(self)
+        return self.recurring_invoice_line_ids.filtered(
+            lambda l: not l.is_canceled and l.recurring_next_date <= date_ref)
+
     @api.multi
     def recurring_create_invoice(self):
-        return self.env[
-            'account.analytic.invoice.line'
-        ].recurring_create_invoice(self)
+        invoice_model = self.env['account.invoice']
+        invoices_values = []
+        for contract in self:
+            contract_lines = contract._get_lines_to_invoice()
```

Comment:
Why looping twice on recurring_invoice_line_ids? you can do something like this
```python
invoice_lines = []
for l in contract.recurring_invoice_line_ids:
    if not l.is_canceled and l.recurring_next_date <= date_ref:
         invoice_lines.append((0, 0, l._prepare_invoice_line(False)))
if invoice_lines:
    invoice_values = contract._prepare_invoice(
                contract.recurring_next_date)
   invoice_values['invoice_line_ids'] = invoice_lines

```

Reply:
Don't you think it's better to keep the definition of a 'line to invoice' in a separate and overridable method ? If we put this logic in `recurring_create_invoice`, we lost the possibility to redefine an invoiceable line.

Plus, I'd like to reduce the logic volume in the `recurring_create_invoice` method because we'll need to replace it by our own grouping-method in PRG (and introduce some code duplication by the way).

Reply:
yes, you are right about overridable method, I will rather add a computed field is_invoiceable that encapsulate the logic of invoiceable lines and treat in the for loop lines with True value.

Still, a minor improvement if the two loops above are not  consequently harmful to performance.

I merege this PR as it is and we will see later if this improvement is needed